### PR TITLE
Support zstd compression for shuffle data

### PR DESCRIPTION
**Pull request by hiboyang**
Issue https://github.com/uber/RemoteShuffleService/issues/90 mentioned zstd has much better compression rate.

Add a conf `spark.shuffle.rss.compress` to allow lz4 and zstd compression.

### CODE REVIEW

File: src/main/scala/org/apache/spark/shuffle/RssShuffleWriter.scala [hunk 1]
```diff
@@ -201,8 +208,19 @@ class RssShuffleWriter[K, V, C](
 
   private def createDataBlock(buffer: Array[Byte]): ByteBuffer = {
     val uncompressedByteCount = buffer.size
-    val compressedBuffer = new Array[Byte](compressor.maxCompressedLength(uncompressedByteCount))
-    val compressedByteCount = compressor.compress(buffer, compressedBuffer)
+    var compressedBuffer: Array[Byte] = null
+    var compressedByteCount: Int = 0
+    if (Compression.COMPRESSION_CODEC_ZSTD.equals(compression)) {
+      compressedBuffer = new Array[Byte](uncompressedByteCount)
+      val n = Zstd.compress(compressedBuffer, buffer, 3)
```
Comment (YutingWang98) [inline src/main/scala/org/apache/spark/shuffle/RssShuffleWriter.scala:215]: spark external shuffle service is using level 1. Would it be better to not hard code it, and make it an RssOpts instead?

Comment (hiboyang) [reply src/main/scala/org/apache/spark/shuffle/RssShuffleWriter.scala]: Good suggestion! Added conf `spark.shuffle.rss.compress.zstd.level`

File: src/main/scala/org/apache/spark/shuffle/rss/BlockDownloaderPartitionRecordIterator.scala [hunk 2]
```diff
@@ -164,12 +170,20 @@ class BlockDownloaderPartitionRecordIterator[K, C](
     val compressedLen = ByteBufUtils.readInt(bytes, 0)
     val uncompressedLen = ByteBufUtils.readInt(bytes, Integer.BYTES)
     val uncompressedBytes = new Array[Byte](uncompressedLen)
-    val count = decompressor.decompress(bytes, Integer.BYTES + Integer.BYTES, uncompressedBytes, 0, uncompressedLen)
-    decompressTime += (System.nanoTime() - decompressStartTime)
-    if (count != compressedLen) {
-      throw new RssInvalidDataException(
-        s"Data corrupted for shuffle $shuffleId partition $partition, expected compressed length: $compressedLen, but it is: $count, " + String.valueOf(downloader))
+    if (Compression.COMPRESSION_CODEC_ZSTD.equals(decompression)) {
+      val n = Zstd.decompress(uncompressedBytes, bytes)
```
Comment (YutingWang98) [inline src/main/scala/org/apache/spark/shuffle/rss/BlockDownloaderPartitionRecordIterator.scala:174]: I think the bytes also inclued uncompressedLen and compressedLen. Better to use `Zstd.decompressByteArray(uncompressedBytes, 0, uncompressedLen, bytes, byteOffset, bytesize - byteOffset)` ? (byteOffset=Integer.BYTES + Integer.BYTES)

Comment (hiboyang) [reply src/main/scala/org/apache/spark/shuffle/rss/BlockDownloaderPartitionRecordIterator.scala:174]: Good catch, you are right! But I did not find method on Zstd to decompress source byte array with given offset. Where did you find method Zstd.decompressByteArray?

Comment (YutingWang98) [reply src/main/scala/org/apache/spark/shuffle/rss/BlockDownloaderPartitionRecordIterator.scala:174]: https://github.com/luben/zstd-jni/blob/master/src/main/java/com/github/luben/zstd/Zstd.java#L406 I saw it from here.

Comment (hiboyang) [reply src/main/scala/org/apache/spark/shuffle/rss/BlockDownloaderPartitionRecordIterator.scala]: Got it, looks like Spark 2.4 uses an old version of Zstd which does not have `Zstd.decompressByteArray` method. I updated the code here a little bit to copy data in another byte array. Could optimize that for higher version of Spark using `Zstd.decompressByteArray` later.

Comment (YutingWang98) [reply src/main/scala/org/apache/spark/shuffle/rss/BlockDownloaderPartitionRecordIterator.scala]: great, thank you!!

**Comment by mabansal** [comment]
@hiboyang can you also please add few tests

**Comment by YutingWang98** [comment]
> 

Sure, will add test to it, thanks

**Comment by hiboyang** [comment]
> @hiboyang can you also please add few tests

@mabansal added test for zstd compression

**Pull request merged**